### PR TITLE
Fix on backpopulate logics - init flow

### PR DESCRIPTION
**Pull request by r1viollet**
# What does this PR do?

- Allow a retry when attaching a mod with a backpopulate when attaching to a PID in the init flow
- Retry with a backpopulate regardless of the failure in the attachement.

# Motivation

Ensure a consistent behaviour between flows.

# Additional Notes

This should not occur often. Only in case of failures to attach.
Risks: In full host, our backpopulate rates are already very high. Could this make things worse ?

# How to test the change?

I do not have a specific test to cover this flow.
This would happen when mappings are ambiguous.

### CODE REVIEW

File: src/unwind_dwfl.cc [hunk 1]
```diff
@@ -248,66 +247,77 @@ DDRes add_runtime_symbol_frame(UnwindState *us, const Dso &dso, ElfAddress_t pc,
...
-
-    DsoHdr::DsoMap const &map = us->dso_hdr.get_pid_mapping(us->pid)._map;
-    if (map.empty()) {
-      int nb_elts;
-      us->dso_hdr.pid_backpopulate(us->pid, nb_elts);
-    }
+  if (us->_dwfl_wrapper->_attached) {
+    return {};
+  }
+  bool backpopulate_attempt = false;
```
Comment (nsavoire) [inline src/unwind_dwfl.cc:290]: I would rename `backpopulate_attempt` to `backpopulate_done` to make it clearer.

File: src/unwind_dwfl.cc [hunk 2]
```diff
@@ -248,66 +247,77 @@ DDRes add_runtime_symbol_frame(UnwindState *us, const Dso &dso, ElfAddress_t pc,
 
   return add_frame(symbol_idx, map_idx, pc, us);
 }
+
+bool try_attach_module(UnwindState *us, const DsoHdr::DsoMap &map,
+                       bool &backpopulate_attempt) {
```
Comment (nsavoire) [inline src/unwind_dwfl.cc:252]: Why pass `backpopulate_attempt` by reference ?

File: src/unwind_dwfl.cc [hunk 3]
```diff
@@ -248,66 +247,77 @@ DDRes add_runtime_symbol_frame(UnwindState *us, const Dso &dso, ElfAddress_t pc,
...
+        us->dso_hdr.get_file_info_value(file_info_id);
+    DDProfMod *ddprof_mod = nullptr;
+    auto res = us->_dwfl_wrapper->register_mod(us->current_ip,
+                                               DsoHdr::get_elf_range(map, it),
+                                               file_info_value, &ddprof_mod);
+    if (IsDDResOK(res)) {
+      success = true;
+      break;
+    }
+    if (!backpopulate_attempt) {
```
Comment (nsavoire) [inline src/unwind_dwfl.cc:274]: Maybe add a comment to explain the logic there

File: src/unwind_dwfl.cc [hunk 4]
```diff
@@ -248,66 +247,77 @@ DDRes add_runtime_symbol_frame(UnwindState *us, const Dso &dso, ElfAddress_t pc,
...
-    }
-    if (!success) {
-      LG_DBG("Unable to attach a mod for PID%d", us->pid);
-      return ddres_warn(DD_WHAT_UW_ERROR);
+  bool success = try_attach_module(us, map, backpopulate_attempt);
+  if (!success && !backpopulate_attempt) {
+    int nb_elts_added;
+    if (us->dso_hdr.pid_backpopulate(us->pid, nb_elts_added) &&
+        nb_elts_added > 0) {
+      success = try_attach_module(us, map, backpopulate_attempt);
```
Comment (nsavoire) [inline src/unwind_dwfl.cc:303]: `backpopulate_attempt = true` probably missing or call `try_attach_module` directly with `true`:
```
success = try_attach_module(us, map, true);
```

**Comment by r1viollet** [comment]
@nsavoire Thanks for the quick review and spotting the issue. I had initially slightly rushed that commit.
Things should look better.

**Review by nsavoire — APPROVED** [review]
LGTM !

**Pull request merged**